source/portable/NetworkInterface/RM46L852/NetworkInterface.c

```diff
@@ -0,0 +1,369 @@
+ */
+
+#include "FreeRTOS.h"
+#include "list.h"
+#include "queue.h"
+#include "semphr.h"
+#include "task.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
+#include "FreeRTOS_IP_Private.h"
+#include "NetworkBufferManagement.h"
+#include "NetworkInterface.h"
+
+#include "emac.h"
+#include "mdio.h"
+#include "phy_dp83640.h"
+#include "sci.h"
+
+
+#if ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES == 0
+    #define ipCONSIDER_FRAME_FOR_PROCESSING( pucEthernetBuffer )    eConsiderFrameForProcessing( ( pucEthernetBuffer ) )
+#else
+    #define ipCONSIDER_FRAME_FOR_PROCESSING( pucEthernetBuffer )    eProcessBuffer
+#endif
+
+
+TaskHandle_t receiveTaskHandle = NULL;
+
+void prvEMACHandlerTask( void * pvParameters  );
+void prvProcessFrame( NetworkBufferDescriptor_t * pxBufferDescriptor );
+uint8   emacAddress[6U] =   {0x00U, 0x08U, 0xEEU, 0x03U, 0xA6U, 0x6CU};
+
+/*
+ * This function is used to call EMAC API to perform low level hardware
+ * initializations.
+ * Also, creates prvEMACHandlerTask for handling received packets.
+ */
+
+BaseType_t xNetworkInterfaceInitialise( void )
+{
+    BaseType_t xFirstCall = pdTRUE;
+    BaseType_t xTaskCreated;
+
+    if (EMACHWInit(emacAddress) == EMAC_ERR_OK)
+    {
+        if( ( xFirstCall == pdTRUE ) || ( receiveTaskHandle == NULL ) )
+        {
+            /* The handler task is created at the highest possible priority to
+             * ensure the interrupt handler can return directly to it. */
+            xTaskCreated = xTaskCreate( prvEMACHandlerTask,
+                                        "EMAC-Handler",
+                                        configMINIMAL_STACK_SIZE,
+                                        NULL,
+                                        configMAX_PRIORITIES - 1,
+                                        &receiveTaskHandle );
+
+            if( ( receiveTaskHandle == NULL ) || ( xTaskCreated != pdPASS ) )
+            {
+                FreeRTOS_printf( ( "Failed to create the handler task." ) );
+
+            }
+            /* After this, the task should not be created. */
+            else
+                xFirstCall = pdFALSE;
+        }
+
+        return pdPASS;
+    }
+    else
+    {
+        return pdFAIL;
+    }
+}
+
+/*
+ *  This function is used to call EMAC API to transmit packets.
+ *  And freeing the network buffers.
+ */
+
+BaseType_t xNetworkInterfaceOutput( NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                    BaseType_t xReleaseAfterSend )
+{
+      struct pbuf_struct xpbuf;
+      xpbuf.payload = pxNetworkBuffer->pucEthernetBuffer;
+      xpbuf.len = pxNetworkBuffer->xDataLength;
+      xpbuf.tot_len = pxNetworkBuffer->xDataLength;
+      xpbuf.next = NULL;
+      if(xpbuf.tot_len < MIN_PKT_LEN)
+      {
+          xpbuf.tot_len = MIN_PKT_LEN;
+          xpbuf.len = MIN_PKT_LEN;
+      }
+
+      taskENTER_CRITICAL();
```

Comment:
At time of writing only 4 Network Interfaces in [FreeRTOS-Plus-TCP/source/portable/NetworkInterface](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/tree/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface) have calls to enter or exit critical sections
The [STM32Hxx/NetworkInterface.c](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/STM32Hxx/NetworkInterface.c#L995) interface calls it exclusively to clear option bits, which is a few assembly instructions.
The other two interfaces, the [Zynq/x_emacspif_hw.c:HandleTxErrors()](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/Zynqx_emacpsif_hw.c#L235) and the [xilinx_ultrascale/x_emacpsif_hw.c:HandleTxErrors()]  **which actually both have the calls commented out.** Where it appears to be there for people to uncomment if attempting to de-bug an issue with a transmit error. 
Lastly the [ThirdParty/MSP432/NetworkInterface.c](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/ThirdParty/MSP432/NetworkInterface.c#L788) which does enter and exit critical sections for its transfers. However, it does this in the task code that executes the packet transfers, not directly in the interface input and output calls. Additionally, this interface has a disclaimer about its use being untested with the current version of the FreeRTOS-Plus-TCP Stack.

**Why does this interface require the use of critical sections to transmit data when no other supported interfaces do?** This approach seems extremely heavy handed as it requires the end user of this interface to essentially set the networking actions as the single most import activity on the system.